### ISSUE TITLE
Split up `decls.nim`

### ISSUE THREAD

**Issue by arnetheduck** [status-im/nim-bearssl#8]
Currently, when importing decls, _all_ bearssl code gets compiled even when parts of the library are not used - it would be good to split it up a little into groups, so that imports can happen in pieces:

RSA, AES, chacha, etc

this way, a user of bearssl will not have to compile everything every time
this task would involve moving the `{.compile.}` statements into different, logically grouped modules - these modules would have to import their dependencies correctly as well.

**Comment by stefantalpalaru** [comment]
Or you could just build a static library, like upstream intended.

**Comment by arnetheduck** [comment]
even with a static library, you need to build `c` files that you don't really need to build, and you run into trouble managing compiler options, versions, debugging information and a number of other inconsistencies between the makefile build environment and the nim build environment. 

Two build systems is worse than one build system, generally.

**Comment by stefantalpalaru** [comment]
> Two build systems is worse than one build system, generally.

Exactly, so stop trying to use the Nim compiler as a build system. It's not fit for the job and it will probably never be good enough at something it's not designed for.

> even with a static library, you need to build c files that you don't really need to build

Only once.

> you run into trouble managing compiler options, versions, debugging information and a number of other inconsistencies between the makefile build environment and the nim build environment

All easier to manage than trying to reinvent a build system using Nim pragmas and then wondering if modules should really set global CFLAGS or if upgrading the wrapped library is too much of a trouble, now that you maintain a parallel list of files to be compiled by the wrapper - some of which not even meant to be part of the library.

**Comment by arnetheduck** [comment]
> Nim compiler as a build system

well, it _is_ a build system: it manages dependencies and rebuilds, it manages complier flags and it managers different versions of compilers. Undeniably, if it works well enough to compile simple C files like here, it is _easier_ to use for Nim developers as it will require only one set of tooling: the same C and Nim compilers that are used to compile Nim code. Many have pointed this out before, it can't be ignored. Some things it will also manage _better_ than makefiles: it understands Nim dependencies and can selectively rebuild the right things in a way that `make` can't.

If we're going to stick to upstream, we should be using `nmake` for windows - somehow, I don't see that happening soon either.

If we want to avoid rebuilds, `ccache` provides a robust enough solution that works per C file and will hash all the right flags, compiler versions, source code, system header file state etc such that there's no risk of getting out-of-date builds because of stale libraries lying around.

**Comment by stefantalpalaru** [comment]
> it is *easier* to use for Nim developers

So the workflows we envision for Nim are ones in which people wrapping C libraries maintain duplicate lists of C files, in various groups to match various Nim modules, then use Nim pragmas and Nim compiler flags to compile them each time wrapper modules are used, then use ccache to mitigate the overhead of compiling the same C files multiple times?

And that's easier than just using the upstream build system to build libraries and link to them from the wrapper?

> simple C files

The ones that had such a confusing compile-time CPU/platform/feature detection that Nim developers saw fit to also duplicate that in Nim code? Everything is "simple" until you look closer.

**Comment by arnetheduck** [comment]
> people wrapping C libraries maintain duplicate lists of C files

**Comment by stefantalpalaru** [comment]
> they make everything in their power so that if people that have installed nim and its deps (make is not one of them) can use the module without any further complications

So "make" is bad, but "ccache" is good?

Is updating the wrapped library such a complication? Because, by fighting upstream and replacing its build system, you make that much harder.

BTW, why are we manually converting .pem certs to C headers during BearSSL updates? https://github.com/status-im/nim-bearssl/commit/8a9503b6dff59985f7f0f5845014fffa557bdec3

**Comment by arnetheduck** [comment]
> So "make" is bad, but "ccache" is good?

I don't recommend using it - it's just one automatic, unobtrusive solution to a particular problem - this issue proposes another, more invasive one: take a big thing and split it to pieces, for the case you don't need anything.

> Is updating the wrapped library such a complication? Because, by fighting upstream and replacing its build system, you make that much harder.

There's an asymmetry here: we update these libraries what... 2-3 times per year? we use them a lot more, so it makes sense to optimize for ease of use.

my preferred solution would actually be that upstream adopted what sqlite and many other "embedding-friendly" libraries do: amalgamation builds. A project like sqlite which is optimized for inclusion in other projects will take this approach and not force a particular build system down anyone's throat. It can be used from Nim, rust, MSBuild, CMake, makefiles and 100 others with equal ease. 

bearssl does not have an advanced build system - just a very verbose one - every file is one function basically - but once you've done the translation once, it's really really easy to keep up to date - I mean, evidently, someone did the initial work for bearssl and didn't consider it too much of a burden.

> BTW, why are we manually converting .pem certs to C headers during BearSSL updates?

my guess would be that this way, the application does not need to load an external file, which again makes the library easier to use (though revocation then means a new application binary instead of a new pem file - one can have arguments about that, I'm sure)

**Comment by stefantalpalaru** [comment]
> amalgamation builds

Terrible idea. If you want whole-program optimisation or link-time optimisation, just use the right compiler flag, instead of concatenating C files like a barbarian: https://en.wikipedia.org/wiki/Interprocedural_optimization

**Comment by arnetheduck** [comment]
you're saying sqlite, one of the most widely deployed pieces of software in the whole wide world, got this wrong and is now suffering for it?

besides, I've been trying to convince you that `lto` is superior in almost all ways to `static archives` for ages, what happened? `ranlib` is truly where the barbarians are gathering :)

**Comment by stefantalpalaru** [comment]
> lto is superior in almost all ways to static archives

Apples and oranges. You're not supposed to use LTO to get rid of unused objects you linked by mistake. You're supposed to to use it for interprocedural optimisation.

> ranlib is truly where the barbarians are gathering :)

You take that back! "ar" and "ranlib" are beautiful tools handed to us by the UNIX gods themselves and we should use them with utmost reverence - preferably from the safe distance of an ancient build system that's aware of the holy pitfalls.

**Comment by yglukhov** [comment]
2 cents. I'm happy i can import bearssl in a nim file and it just works when compiling to android and ios and I don't have to adjust my build system in any way. Otherwise bearssl would not have any benefit over openssl for me.

**Comment by arnetheduck** [comment]
#27
